### ISSUE TITLE
Any document about how dxdp calculated in your code?

### ISSUE THREAD

**Issue by lvzhaoyang** [MightyChaos/LKVOLearner#2]
Thanks for the nice work. I was reading the code to better understand the paper. But I don't understand about this line:

https://github.com/MightyChaos/LKVOLearner/blob/4947803d7e722b9c434f5fc3b0068cd88c08612e/src/DirectVOLayer.py#L330

If I understand it correctly, it should be the Jacobian of x with respect to p, where x is the pixel location in the image space, and p is the lie vector for the transform in se3. I tried to derive the Jacobian to get the same form as your code, but I have not figured it out yet. Is this part illustrated anywhere in supplementary? Can you share your derivation for this formula, or refer me to any document?

**Comment by MightyChaos** [comment]
I did it myself and didn't include in the supplementary.  I can give you a brief outline of the derivation here, hope it helps.

first, $p$ is a 6-dim vector with the first 3 as the exponential map coordinate $w$, and the last 3 as the translation vector $\mathbf{t}$. and the pixel coordinate is denoted as (x,y)
Recall the warping function is written as:
$$[x', y']^T = <R[x,y,1]^T+d\mathbf{t}>$$, where $d$ is the inverse depth, and <.> denotes projection.
what we try to do is to compute the jacobian at the identity pose $p_0$, where $R$ is simply an identity matrix, and $t$ is zero vector. 
To compute this, you use chain rule, i.e., first compute the Jacobian of the projection function $<[x,y,z]^T>=[x/z, y/z]^T$, then multiply it by the derivative for the rigid transformation inside the projection function.
An important fact to know is that $R = \exp^{[w]_\times}$, and it can be simplified as $R = I + [w]_\times$ when $w$ is around zero. (which is known as the small angle approximation)

With this, the rest of the derivation should be straightforward.

**Comment by lvzhaoyang** [comment]
Thanks for the explaination. Now I figured it out. :) 

I did not use your approximation, but directly operate on the lie vector, but it gets the same close-form result. I just add the derivation here for a bookmark. 
The derivative of pose vector in se3 in the origin (with your inverse depth parameterization): 
![image](https://user-images.githubusercontent.com/3089615/42931509-2b2cd720-8b40-11e8-925d-7434c268be4e.png)
And the derivative of the projection matrix, with the value [x'=x, y'=y, z'=1]: 
![image](https://user-images.githubusercontent.com/3089615/42932248-158b406c-8b42-11e8-858b-7c2de478d5db.png)

With the chain rule I get your solution:
![image](https://user-images.githubusercontent.com/3089615/42932428-84cbb66e-8b42-11e8-8c37-ddc7eda2246d.png)

I also just noticed that your premultiply the focal length to the image gradient. Looks good. 

https://github.com/MightyChaos/LKVOLearner/blob/4947803d7e722b9c434f5fc3b0068cd88c08612e/src/DirectVOLayer.py#L287
